raft/src/test/java/org/apache/kafka/raft/LeaderStateTest.java

```diff
@@ -103,189 +105,247 @@ public void testNonFollowerAcknowledgement() {
+//        assertEquals(caughtUpTime, describeObserverState(state, node1, currentTime).lastCaughtUpTimestamp());
+//    }
+//
+//    @Test
+//    public void testIdempotentEndOffsetUpdate() {
+//        LeaderState<?> state = newLeaderState(singleton(localId), 15L);
+//        assertEquals(Optional.empty(), state.highWatermark());
+//        assertTrue(state.updateLocalState(new LogOffsetMetadata(16L)));
+//        assertFalse(state.updateLocalState(new LogOffsetMetadata(16L)));
+//        assertEquals(Optional.of(new LogOffsetMetadata(16L)), state.highWatermark());
+//    }
+//
+//    @Test
+//    public void testUpdateHighWatermarkMetadata() {
+//        LeaderState<?> state = newLeaderState(singleton(localId), 15L);
+//        assertEquals(Optional.empty(), state.highWatermark());
+//
+//        LogOffsetMetadata initialHw = new LogOffsetMetadata(16L, Optional.of(new MockOffsetMetadata("bar")));
+//        assertTrue(state.updateLocalState(initialHw));
+//        assertEquals(Optional.of(initialHw), state.highWatermark());
+//
+//        LogOffsetMetadata updateHw = new LogOffsetMetadata(16L, Optional.of(new MockOffsetMetadata("baz")));
+//        assertTrue(state.updateLocalState(updateHw));
+//        assertEquals(Optional.of(updateHw), state.highWatermark());
+//    }
+//
+//    @Test
+//    public void testUpdateHighWatermarkQuorumSizeTwo() {
+//        int otherNodeId = 1;
+//        LeaderState<?> state = newLeaderState(mkSet(localId, otherNodeId), 10L);
+//        assertFalse(state.updateLocalState(new LogOffsetMetadata(13L)));
+//        assertEquals(singleton(otherNodeId), state.nonAcknowledgingVoters());
+//        assertEquals(Optional.empty(), state.highWatermark());
+//        assertFalse(state.updateReplicaState(otherNodeId, 0, new LogOffsetMetadata(10L)));
+//        assertEquals(emptySet(), state.nonAcknowledgingVoters());
+//        assertEquals(Optional.empty(), state.highWatermark());
+//        assertTrue(state.updateReplicaState(otherNodeId, 0, new LogOffsetMetadata(11L)));
+//        assertEquals(Optional.of(new LogOffsetMetadata(11L)), state.highWatermark());
+//        assertTrue(state.updateReplicaState(otherNodeId, 0, new LogOffsetMetadata(13L)));
+//        assertEquals(Optional.of(new LogOffsetMetadata(13L)), state.highWatermark());
+//    }
 
     @Test
-    public void testNonMonotonicLocalEndOffsetUpdate() {
-        LeaderState<?> state = newLeaderState(singleton(localId), 15L);
-        assertEquals(Optional.empty(), state.highWatermark());
-        assertTrue(state.updateLocalState(new LogOffsetMetadata(16L)));
-        assertEquals(Optional.of(new LogOffsetMetadata(16L)), state.highWatermark());
-        assertThrows(IllegalStateException.class,
-            () -> state.updateLocalState(new LogOffsetMetadata(15L)));
-    }
-
-    @Test
-    public void testLastCaughtUpTimeVoters() {
+    public void testUpdateHighWatermarkQuorumSizeThree() {
```

Comment:
git is botching the diffs. the only changes to this test file are commenting out tests (temporary) and _adding_ new tests `testUpdateHighWatermarkRemovingFollowerFromVoterStates` and `testUpdateHighWatermarkRemovingLeaderFromVoterStates`